node/cn/tracers/api.go

```diff
@@ -451,6 +455,9 @@ func (api *CommonAPI) traceChain(start, end *types.Block, config *TraceConfig, n
 				failed = err
 				break
 			}
+			// Insert parent hash in history contract.
+			api.backend.Engine().Initialize(newChainContext(localctx, api.backend), next.Header(), statedb)
```

Comment:
Is this necesesary here? I think this is called by the call chain:
```
api.go:453 StateAtBlock()
api_backend.go:404 stateAtBlock()
state_accessor.go:167 Process()
```

Reply:
I think `Process()` might not be touched. Thus, marking the blockhash at this point as well even if the `readOnly` is false.

Reply:
`StateAtBlock` seems to only return statedb only when the "correct" one is generated.
If `Initialize` wasn't called by `Process`, then `Finalize` would not be called as well, in which case `StateAtBlock` will return an error.

btw, is `Initialize` idempotent? (I think it is now, but perhaps can change later)

Reply:
> If Initialize wasn't called by Process, then Finalize would not be called as well

The `Finalize` of state database(not engine) is done in `ProcessParentBlockHash`. Does it answer indirectly?

> btw, is Initialize idempotent? (I think it is now, but perhaps can change later)

If you mean the idempotent in a unit of single block, then yes. I'm sure you were asking in this context.
Of course, calling two times with the same parameter at different block numbers(e.g., `n` and `n+1`) differs, which stores the same parameter(hash) into `block.number - 1` and `block.number`, respectively.

Reply:
I'm not talking about `statedb.Finalise` (with an "s").
`Initialize` always goes in pair with `engine.Finalize`, right? (please correct me if I'm wrong)
`Process` either (1) calls both `Initialize` and `Finalize`, or (2) calls neither.
`StateAtBlock` does not return an error only in the first case.

I'm concerning if later in the future when `Initialize` stops being idempotent, this line's additional call can be problematic.
Maybe I'm over concerning, so please tell me your opinion.

Also, I just found out if we're going to leave this code, I think it should go to the bottom (after statedb is sent to the channel):
https://github.com/kaiachain/kaia/blob/8fb6105b4c0455caefc67caf7b314b40b7333912/node/cn/tracers/api.go#L469-L478

Reply:
I wanted all Initialize from `api.go` be removed, but I didn't think we could.
After reading ollie's comment, I took another look, and I think we can.

Reply:
Sorry, I didn't know it was for tracing N+1 block where StateAtBlock(N) + Initialize is required.